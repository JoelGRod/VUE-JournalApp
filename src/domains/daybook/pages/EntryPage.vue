<template>
  <template v-if="entry">
    <div class="entry-title d-flex justify-content-between p-2">
      <div>
        <span class="text-success fs-3 fw-bold">{{ getDay }}</span>
        <span class="mx-1 fs-3">{{ getMonth }}</span>
        <span class="mx-2 fs-4 fw-light">{{ getYearDay }}</span>
      </div>

      <div>
        <button class="btn btn-danger mx-2"
          @click="removeEntry()">
          Delete
          <i class="fa fa-trash-alt"></i>
        </button>
        <button class="btn btn-primary">
          Upload Image
          <i class="fa fa-upload"></i>
        </button>
      </div>
    </div>

    <hr />
    <div class="d-flex flex-column px-3 h-75">
      <textarea
        class="entry-textarea"
        placeholder="What's happen?"
        v-model="entry.text"
      ></textarea>
    </div>

    <img
      src="https://images.pexels.com/photos/4245826/pexels-photo-4245826.jpeg?cs=srgb&dl=pexels-riccardo-bertolo-4245826.jpg&fm=jpg"
      alt="entry-photo"
      class="img-thumbnail entry-img"
    />
  </template>

  <FabComponent icon="fa-save" @on:click="saveEntry()"></FabComponent>
  
</template>

<script>
import { defineAsyncComponent } from "@vue/runtime-core";
import { mapGetters, mapActions } from "vuex";
import getDayMonthYear from "../helpers/getDayMonthYear";

export default {
  name: "entry-page",
  props: {
    entryId: {
      type: String,
      required: true,
    },
  },
  data() {
    return {
      entry: null,
    };
  },
  methods: {
    ...mapActions('daybook', 
      ['updateEntry', 'createEntry', 'deleteEntry']
    ),
    loadEntry() {
      let entry = {}

      if( this.entryId === 'new' ) {
        entry = {
          date: new Date().getTime(),
          text: ''
        }
      } else {
        entry = this.getEntryById(this.entryId);
        if (entry === undefined)
          return this.$router.push({ name: "Daybook-No-Entry" });
      }

      this.entry = entry;
    },
    async saveEntry() {
      if( this.entry.id ) {
        await this.updateEntry(this.entry)
        // Confirmation alert
      } else {
        const id = await this.createEntry(this.entry)
        this.$router.push({ name: 'Daybook-Entry', params: { id } })
      }
    },
    async removeEntry() {
      if(this.entryId === 'new') return;
      await this.deleteEntry(this.entryId)
      // Navigate to no entry selected
    }
  },
  computed: {
    ...mapGetters("daybook", ["getEntryById"]),
    getDay() {
      const { day } = getDayMonthYear(this.entry.date);
      return day;
    },
    getMonth() {
      const { month } = getDayMonthYear(this.entry.date);
      return month;
    },
    getYearDay() {
      const { yearDay } = getDayMonthYear(this.entry.date);
      return yearDay;
    },
  },
  watch: {
    entryId() {
      this.loadEntry();
    },
  },
  components: {
    FabComponent: defineAsyncComponent(() =>
      import("../components/FabComponent")
    ),
  },
  created() {
    this.loadEntry();
  },
};
</script>

<style lang="scss" scoped>
.entry-textarea {
  font-size: 1.2rem;
  border: none;
  height: 100%;

  &:focus {
    outline: none;
  }
}

.entry-img {
  position: fixed;
  bottom: 9rem;
  right: 1rem;
  width: 12rem;
  box-shadow: 0px 5px 10px rgba($color: #333333, $alpha: 0.2);
}
</style>