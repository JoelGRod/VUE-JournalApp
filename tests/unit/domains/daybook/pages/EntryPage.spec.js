import { shallowMount } from "@vue/test-utils";
import { createStore } from "vuex";

import EntryPage from "@/domains/daybook/pages/EntryPage";

import daybookStore from "@/domains/daybook/store/daybook";
import { daybookState } from '../../../mock-data/test-daybook-store';

describe('Testing: EntryPage.vue', () => {

    const createVuexStore = (initialState) =>
        createStore({
            modules: {
                daybook: {
                    ...daybookStore,
                    state() {
                        return { ...initialState }
                    }
                }
            }
        });
    const store = createVuexStore(daybookState);

    const routerMock = {
        push: jest.fn()
    }

    let wrapper;
    beforeEach(() => {
        jest.clearAllMocks();
        wrapper = shallowMount(EntryPage, {
            global: {
                mocks: {
                    $router: routerMock
                },
                plugins: [store]
            },
            props: {
                entryId: daybookState.entries[0].id
            }
        });
    })

    test('should match with the snapshot', () => {
        // Assert
        expect( wrapper.html() ).toMatchSnapshot();
    });

    test('should redirect user if id is not found in state', () => {
        // Arrange
        const wrapper = shallowMount(EntryPage, {
            global: {
                mocks: {
                    $router: routerMock
                },
                plugins: [store]
            },
            props: {
                entryId: 'Not a valid id'
            }
        });
        // Act
        // Assert
        expect(routerMock.push).toHaveBeenCalled();
        expect(routerMock.push)
            .toHaveBeenCalledWith(
                { name: "Daybook-No-Entry" }
            );
    });

    test('should show the entry defined in props correctly', () => {
        // Assert
        expect( routerMock.push ).not.toHaveBeenCalled();
    });
});