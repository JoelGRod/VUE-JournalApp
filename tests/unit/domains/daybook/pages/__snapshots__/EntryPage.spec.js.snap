// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing: EntryPage.vue should match with the snapshot 1`] = `
<div class="entry-title d-flex justify-content-between p-2">
  <div><span class="text-success fs-3 fw-bold">4</span><span class="mx-1 fs-3">Septiembre</span><span class="mx-2 fs-4 fw-light">2021, SÃ¡bado</span></div>
  <div><input type="file" multiple="" accept="image/png, image/jpeg" style="display: none;"><button class="btn btn-danger mx-2"> Delete <i class="fa fa-trash-alt"></i></button><button class="btn btn-primary"> Upload Image <i class="fa fa-upload"></i></button></div>
</div>
<hr>
<div class="d-flex flex-column px-3 h-75"><textarea class="entry-textarea" placeholder="What's happen?"></textarea></div><img src="https://res.cloudinary.com/do7c3iy3j/image/upload/v1630758299/vue-tests/odv8p7rkeuiqrslqf6ff.jpg" alt="entry-photo" class="img-thumbnail entry-img">
<!--v-if-->
<fab-component-stub icon="fa-save"></fab-component-stub>
`;
